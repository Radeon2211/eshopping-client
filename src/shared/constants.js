import * as Yup from 'yup';
import { getCountries } from 'country-fns';

export const anyStringRule = Yup.string().min(1).required();

export const userRules = {
  email: Yup.string().email().max(320).trim().required(),
  hideEmail: Yup.bool(),
  username: Yup.string().min(3).max(20).trim().required(),
  password: Yup.string().min(7).max(64).trim().required(),
  firstName: Yup.string().max(60).trim().required(),
  lastName: Yup.string().max(80).trim().required(),
  phonePrefix: Yup.object()
    .shape({
      value: Yup.string().required(),
      label: Yup.string().required(),
    })
    .nullable()
    .required(),
  phoneNumber: Yup.string()
    .min(5)
    .max(15)
    .matches(/^[\d-]{5,15}$/)
    .trim()
    .required(),
  hidePhone: Yup.bool(),
  street: Yup.string().max(60).trim().required(),
  zipCode: Yup.string().max(12).trim().required(),
  city: Yup.string().max(100).trim().required(),
  country: Yup.object()
    .shape({
      value: Yup.string().required(),
      label: Yup.string().required(),
    })
    .nullable()
    .required(),
};

export const productConditions = {
  NEW: 'new',
  USED: 'used',
  NOT_APPLICABLE: 'not_applicable',
};

export const productRules = {
  name: Yup.string().max(150).trim().required(),
  price: Yup.number().min(0.01).max(1000000).required(),
  quantity: Yup.number().min(1).max(100000).required(),
  condition: Yup.string()
    .oneOf([productConditions.NEW, productConditions.USED, productConditions.NOT_APPLICABLE])
    .required(),
  description: Yup.string().max(600).trim(),
};

export const listOfAreaCodes = getCountries().map(({ name, dial }) => {
  const finalValue = `+${dial} ${name.split('(')[0].trim()}`;
  return {
    value: dial,
    label: finalValue,
  };
});

export const listOfCountries = getCountries().map(({ name }) => {
  const finalValue = name.split('(')[0].trim();
  return {
    value: finalValue,
    label: finalValue,
  };
});

export const productPages = {
  ALL_PRODUCTS: 'ALL_PRODUCTS',
  MY_PRODUCTS: 'MY_PRODUCTS',
  USER_PRODUCTS: 'USER_PRODUCTS',
};

export const itemTypes = {
  CART: 'CART',
  ORDER: 'ORDER',
  TRANSACTION: 'TRANSACTION',
};

export const listItemTypes = {
  PRODUCT: 'PRODUCT',
  ORDER: 'ORDER',
};

export const orderTypes = {
  PLACED_ORDERS: 'PLACED_ORDERS',
  SELL_HISTORY: 'SELL_HISTORY',
};

export const inputKinds = {
  INPUT: 'INPUT',
  TEXTAREA: 'TEXTAREA',
  SELECT: 'SELECT',
  RADIO: 'RADIO',
};

export const defaultProductsPerPage = 10;

export const defaultOrdersPerPage = 6;

export const defaultAppPath = '/products?p=1';

export const filtersActions = {
  INIT_STATE: 'INIT_STATE',
  SET_SORT_BY: 'SET_SORT_BY',
  SET_MIN_PRICE: 'SET_MIN_PRICE',
  SET_MAX_PRICE: 'SET_MAX_PRICE',
  SET_CONDITION: 'SET_CONDITION',
};

export const sliderPositionsActions = {
  SET_BOTH: 'SET_BOTH',
  SET_LEFT: 'SET_LEFT',
  SET_RIGHT: 'SET_RIGHT',
};

export const modalTypes = {
  ABOUT_WEBSITE: 'ABOUT_WEBSITE',
  ADD_ADMIN: 'ADD_ADMIN',
  ADD_PRODUCT: 'ADD_PRODUCT',
  BUY_PRODUCTS: 'BUY_PRODUCTS',
  CART_ITEM_ADDED: 'CART_ITEM_ADDED',
  CHANGE_NAME: 'CHANGE_NAME',
  CHANGE_EMAIL: 'CHANGE_EMAIL',
  CHANGE_USERNAME: 'CHANGE_USERNAME',
  CHANGE_PHONE_NUMBER: 'CHANGE_PHONE_NUMBER',
  CHANGE_ADDRESS: 'CHANGE_ADDRESS',
  CHANGE_PASSWORD: 'CHANGE_PASSWORD',
  CHANGE_CONTACTS: 'CHANGE_CONTACTS',
  CHANGE_DELIVERY_ADDRESS: 'CHANGE_DELIVERY_ADDRESS',
  CLEAR_CART: 'CLEAR_CART',
  DELETE_ACCOUNT: 'DELETE_ACCOUNT',
  DELETE_PRODUCT: 'DELETE_PRODUCT',
  EDIT_PRODUCT: 'EDIT_PRODUCT',
  LOGIN: 'LOGIN',
  REMOVE_ADMIN: 'REMOVE_ADMIN',
  RESET_PASSWORD: 'RESET_PASSWORD',
  PENDING_USER_INFO: 'PENDING_USER_INFO',
  SEND_VERIFICATION_LINK: 'SEND_VERIFICATION_LINK',
  SIGNUP: 'SIGNUP',
};

export const sortProductsOptions = [
  { value: undefined, label: 'Default sorting' },
  { value: 'price:asc', label: `Price - ascending` },
  { value: 'price:desc', label: 'Price - descending' },
  { value: 'name:asc', label: 'Name - A to Z' },
  { value: 'name:desc', label: 'Name - Z to A' },
];

export const sortOrdersOptions = [
  { value: 'createdAt:desc', label: 'Date - from newest' },
  { value: 'createdAt:asc', label: 'Date - from oldest' },
  { value: 'overallPrice:asc', label: `Price - ascending` },
  { value: 'overallPrice:desc', label: 'Price - descending' },
];

export const singleInfoNames = {
  USERNAME: 'Username',
  NAME: 'Name',
  EMAIL: 'Email',
  PHONE_NUMBER: 'Phone number',
  ADDRESS: 'Address',
  CONTACTS: 'Contacts',
};

export const updateCartActions = {
  INCREMENT: 'INCREMENT',
  DECREMENT: 'DECREMENT',
  NUMBER: 'NUMBER',
};

export const defaultErrorMessage = 'Something went wrong';

export const productsPerPageControllerOptions = [
  { value: 1, label: '1' },
  { value: 2, label: '2' },
  { value: 3, label: '3' },
  { value: 4, label: '4' },
  { value: 5, label: '5' },
  { value: 10, label: '10' },
  { value: 15, label: '15' },
  { value: 20, label: '20' },
  { value: 25, label: '25' },
];

export const defaultScrollToConfig = {
  top: 0,
  left: 0,
  behavior: 'smooth',
};

export const userStatuses = {
  ACTIVE: 'active',
  PENDING: 'pending',
};
