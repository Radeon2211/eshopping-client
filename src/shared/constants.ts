import * as Yup from 'yup';
import countryCodes from 'country-codes-list';
import { ProductCondition, ProductPhotoFieldValue } from './types/enums';

export const anyStringRule = Yup.string().min(1).required();

export const userRules = {
  email: Yup.string().email().max(320).trim().required(),
  hideEmail: Yup.bool(),
  username: Yup.string().min(3).max(20).trim().required(),
  password: Yup.string().min(7).max(64).trim().required(),
  firstName: Yup.string().max(60).trim().required(),
  lastName: Yup.string().max(80).trim().required(),
  phonePrefix: Yup.object()
    .shape({
      value: Yup.string().required(),
      label: Yup.string().required(),
    })
    .nullable()
    .required(),
  phoneNumber: Yup.string()
    .min(5)
    .max(15)
    .matches(/^[\d-]{5,15}$/)
    .trim()
    .required(),
  hidePhone: Yup.bool(),
  street: Yup.string().max(60).trim().required(),
  zipCode: Yup.string().max(12).trim().required(),
  city: Yup.string().max(100).trim().required(),
  country: Yup.object()
    .shape({
      value: Yup.string().required(),
      label: Yup.string().required(),
    })
    .nullable()
    .required(),
};

export const productRules = {
  name: Yup.string().max(150).trim().required(),
  price: Yup.number().min(0.01).max(1000000).required(),
  quantity: Yup.number().min(1).max(100000).required(),
  condition: Yup.string()
    .oneOf([ProductCondition.NEW, ProductCondition.USED, ProductCondition.NOT_APPLICABLE])
    .required(),
  description: Yup.string().max(600).trim(),
  photo: Yup.mixed()
    .test('is-error', (value) => value !== ProductPhotoFieldValue.ERROR)
    .nullable(),
};

export const listOfAreaCodes = countryCodes.customArray({
  value: '{countryCallingCode}',
  label: '+{countryCallingCode} {countryNameEn}',
});

export const listOfCountries = countryCodes.customArray({
  value: '{countryNameEn}',
  label: '{countryNameEn}',
});

export const itemTypes = {
  CART: 'CART',
  ORDER: 'ORDER',
  TRANSACTION: 'TRANSACTION',
};

export const listItemTypes = {
  PRODUCT: 'PRODUCT',
  ORDER: 'ORDER',
};

export const inputKinds = {
  INPUT: 'INPUT',
  TEXTAREA: 'TEXTAREA',
  SELECT: 'SELECT',
  RADIO: 'RADIO',
};

export const defaultProductsPerPage = 10;

export const defaultOrdersPerPage = 6;

export const defaultAppPath = '/products?p=1';

export const filtersActions = {
  INIT_STATE: 'INIT_STATE',
  SET_SORT_BY: 'SET_SORT_BY',
  SET_MIN_PRICE: 'SET_MIN_PRICE',
  SET_MAX_PRICE: 'SET_MAX_PRICE',
  SET_CONDITION: 'SET_CONDITION',
};

export const sliderPositionsActions = {
  SET_BOTH: 'SET_BOTH',
  SET_LEFT: 'SET_LEFT',
  SET_RIGHT: 'SET_RIGHT',
};

export const sortProductsOptions = [
  { value: undefined, label: 'Default sorting' },
  { value: 'price:asc', label: 'Price - ascending' },
  { value: 'price:desc', label: 'Price - descending' },
  { value: 'name:asc', label: 'Name - A to Z' },
  { value: 'name:desc', label: 'Name - Z to A' },
];

export const sortOrdersOptions = [
  { value: 'createdAt:desc', label: 'Date - from newest' },
  { value: 'createdAt:asc', label: 'Date - from oldest' },
  { value: 'overallPrice:asc', label: 'Price - ascending' },
  { value: 'overallPrice:desc', label: 'Price - descending' },
];

export const singleInfoNames = {
  USERNAME: 'Username',
  NAME: 'Name',
  EMAIL: 'Email',
  PHONE_NUMBER: 'Phone number',
  ADDRESS: 'Address',
  CONTACTS: 'Contacts',
};

export const defaultErrorMessage = 'Something went wrong';

export const productsPerPageControllerOptions = [
  { value: 1, label: '1' },
  { value: 2, label: '2' },
  { value: 3, label: '3' },
  { value: 4, label: '4' },
  { value: 5, label: '5' },
  { value: 10, label: '10' },
  { value: 15, label: '15' },
  { value: 20, label: '20' },
  { value: 25, label: '25' },
];

export const defaultScrollToConfig: ScrollToOptions = {
  top: 0,
  left: 0,
  behavior: 'smooth',
};
