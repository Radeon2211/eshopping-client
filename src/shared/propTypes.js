import PropTypes from 'prop-types';

export const deliveryAddressStrings = {
  firstName: PropTypes.string,
  lastName: PropTypes.string,
  phonePrefix: PropTypes.string,
  phoneNumber: PropTypes.string,
  street: PropTypes.string,
  zipCode: PropTypes.string,
  city: PropTypes.string,
  country: PropTypes.string,
};

export const userProfile = {
  _id: PropTypes.string.isRequired,
  email: PropTypes.string.isRequired,
  username: PropTypes.string.isRequired,
  status: PropTypes.string.isRequired,
  createdAt: PropTypes.string.isRequired,
  contacts: PropTypes.shape({
    email: PropTypes.bool.isRequired,
    phone: PropTypes.bool.isRequired,
  }).isRequired,
  ...deliveryAddressStrings,
};

export const fullProductItem = {
  _id: PropTypes.string.isRequired,
  quantitySold: PropTypes.number.isRequired,
  buyerQuantity: PropTypes.number.isRequired,
  name: PropTypes.string.isRequired,
  price: PropTypes.number.isRequired,
  quantity: PropTypes.number.isRequired,
  description: PropTypes.string.isRequired,
  condition: PropTypes.string.isRequired,
  createdAt: PropTypes.string.isRequired,
  updatedAt: PropTypes.string.isRequired,
  __v: PropTypes.number.isRequired,
  photo: PropTypes.bool.isRequired,
  seller: PropTypes.shape({
    username: PropTypes.string.isRequired,
  }).isRequired,
};

export const cartItem = {
  _id: PropTypes.string.isRequired,
  quantity: PropTypes.number.isRequired,
  product: PropTypes.shape(fullProductItem).isRequired,
};

export const orderItem = {
  _id: PropTypes.string.isRequired,
  products: PropTypes.arrayOf(
    PropTypes.shape({
      _id: PropTypes.string.isRequired,
      name: PropTypes.string.isRequired,
      price: PropTypes.number.isRequired,
      quantity: PropTypes.number.isRequired,
      photo: PropTypes.bool.isRequired,
    }),
  ).isRequired,
  seller: PropTypes.shape({
    username: PropTypes.string.isRequired,
    phone: PropTypes.string,
    email: PropTypes.string,
  }),
  buyer: PropTypes.shape({
    username: PropTypes.string.isRequired,
  }),
  deliveryAddress: PropTypes.shape(deliveryAddressStrings).isRequired,
  overallPrice: PropTypes.number.isRequired,
  createdAt: PropTypes.string.isRequired,
  updatedAt: PropTypes.string.isRequired,
  __v: PropTypes.number.isRequired,
};

export const signupErrors = {
  email: PropTypes.string,
  username: PropTypes.string,
  password: PropTypes.string,
  ...deliveryAddressStrings,
};

export const signupTouched = {
  email: PropTypes.bool,
  username: PropTypes.bool,
  password: PropTypes.bool,
  firstName: PropTypes.bool,
  lastName: PropTypes.bool,
  phonePrefix: PropTypes.oneOfType([PropTypes.bool, PropTypes.object]),
  phoneNumber: PropTypes.bool,
  street: PropTypes.bool,
  zipCode: PropTypes.bool,
  city: PropTypes.bool,
  country: PropTypes.oneOfType([PropTypes.bool, PropTypes.object]),
  'react-select-3-input': PropTypes.bool,
  'react-select-4-input': PropTypes.bool,
};

export const signupValues = {
  email: PropTypes.string.isRequired,
  username: PropTypes.string.isRequired,
  password: PropTypes.string.isRequired,
  firstName: PropTypes.string.isRequired,
  lastName: PropTypes.string.isRequired,
  phonePrefix: PropTypes.shape({
    value: PropTypes.string.isRequired,
    label: PropTypes.string.isRequired,
  }),
  phoneNumber: PropTypes.string.isRequired,
  street: PropTypes.string.isRequired,
  zipCode: PropTypes.string.isRequired,
  city: PropTypes.string.isRequired,
  country: PropTypes.shape({
    value: PropTypes.string.isRequired,
    label: PropTypes.string.isRequired,
  }),
};
