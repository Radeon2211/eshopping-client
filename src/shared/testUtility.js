// eslint-disable-next-line import/no-extraneous-dependencies
import checkPropTypes from 'check-prop-types';

export const checkProps = (component, expectedProps) => {
  // eslint-disable-next-line react/forbid-foreign-prop-types
  return checkPropTypes(component.propTypes, expectedProps, 'props', component.name);
};

export const historyPageNum = (pageNumber) => ({
  listen: jest.fn(),
  location: { search: `?p=${pageNumber}` },
  createHref: jest.fn(),
  push: jest.fn(),
});

export const propsPagination = (itemQuantity = 5) => ({
  itemQuantity,
  isDataLoading: false,
  maxQuantityPerPage: 2,
});
