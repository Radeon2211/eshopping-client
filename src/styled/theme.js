const theme = {
  colors: {
    light1: '#FAFAFA',
    light1Transparent: '#FAFAFA88',
    light2: '#DADFE1',
    light3: '#BFBFBF',
    light4: '#666',
    blue: '#1E8BC3',
    blueLight: '#22A7F0',
    blueDark: '#67809F',
    green: '#00A790',
    greenDark: '#009680',
    red: '#F22613',
  },
  shadows: {
    level1: '0 0 2px rgba(0, 0, 0, .1)',
    level2: '0 0 2px rgba(0, 0, 0, .3)',
    level3: '0 0 3px rgba(0, 0, 0, .3)',
    level4: '0 0 5px rgba(0, 0, 0, .3)',
    level5: '0 0 8px rgba(0, 0, 0, .3)',
  },
  fonts: {
    logo: 'Lobster, cursive',
    heading: 'Ubuntu, sans-serif',
    text: 'Lato, sans-serif',
  },
  spacings: {
    level1: '0.8rem',
    level2: '1.2rem',
    level3: '2.4rem',
    level4: '3.6rem',
    level5: '4.8rem',
  },
  zIndexes: {
    level1: 10,
    level2: 100,
    level3: 1000,
    level4: 10000,
  },
};

export default theme;
