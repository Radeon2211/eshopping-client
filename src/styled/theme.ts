const theme = {
  colors: {
    light1: '#FAFAFA',
    light1Transparent: '#FAFAFA88',
    light2: '#DADFE1',
    light3: '#BFBFBF',
    light4: '#777',
    blue: '#1E8BC3',
    blueLight: '#22A7F0',
    blueDark: '#67809F',
    green: '#06AD96',
    greenDark: '#009680',
    red: '#F22613',
    redLight: '#F94D2A',
    redDark: '#985A5A',
  },
  shadows: {
    level1: '0 0 3px rgba(0, 0, 0, 0.03)',
    level2: '0 0 3px rgba(0, 0, 0, 0.3)',
    level3: '0 0 5px rgba(0, 0, 0, 0.3)',
  },
  fonts: {
    logo: 'Lobster, cursive',
    heading: 'Noto Sans, sans-serif',
    text: 'Nunito, sans-serif',
  },
  spacings: {
    level1: '0.8rem',
    level2: '1.2rem',
    level3: '2.4rem',
    level4: '3.6rem',
    level5: '4.8rem',
  },
  zIndexes: {
    level1: 10,
    level2: 100,
    level3: 1000,
    level4: 10000,
    level5: 100000,
  },
  fontSizes: {
    level1: '1.3rem',
    level2: '1.4rem',
    level3: '1.5rem',
    level4: '1.7rem',
    level5: '2.2rem',
    level6: '2.6rem',
  },
  durations: {
    level1: 0.1,
    level2: 0.2,
    level3: 0.3,
  },
  lineHeights: {
    level1: '1',
    level2: '1.1',
    level3: '1.2',
    level4: '1.3',
    level5: '1.4',
    level6: '1.5',
  },
};

export type ThemeSpacing = keyof typeof theme.spacings;
export type ThemeZIndex = keyof typeof theme.zIndexes;
export type ThemeFontSize = keyof typeof theme.fontSizes;
export type ThemeDuration = keyof typeof theme.durations;
export type ThemeLineHeight = keyof typeof theme.lineHeights;

export default theme;
