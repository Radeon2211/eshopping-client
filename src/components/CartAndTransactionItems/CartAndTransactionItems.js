import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { Link } from 'react-router-dom';
import TransactionAndOrderProdItem from '../TransactionAndOrderProdItem/TransactionAndOrderProdItem';
import CartItem from './CartItem/CartItem';
import PlainText from '../UI/PlainText';
import { GreenText } from '../../styled/components';
import { itemTypes } from '../../shared/constants';

export const SC = {};
SC.SingleSeller = styled.div`
  padding: ${({ theme }) => theme.spacings.level3} 0;

  &:not(:last-child) {
    border-bottom: 1px solid ${({ theme }) => theme.colors.light3};
  }

  &:last-child {
    padding-bottom: 0;
  }
`;

const CartAndTransactionItems = (props) => {
  const { items, type, isCartLoading } = props;

  const sellersObject = items.reduce((acc, item) => {
    const sellerUsername =
      type === itemTypes.CART ? item.product.seller.username : item.seller.username;
    if (!acc[sellerUsername]) {
      acc[sellerUsername] = {
        items: [],
        sellerUsername,
      };
    }
    acc[sellerUsername].items.push(item);
    return acc;
  }, {});

  const sellersArray = Object.entries(sellersObject).map(([sellerUsername, rest]) => ({
    sellerUsername,
    ...rest,
  }));

  const sellerList = sellersArray.map(({ sellerUsername, items: listItem }) => (
    <SC.SingleSeller key={sellerUsername} data-testid="CartAndTransactionItems-item">
      <PlainText size="3" display="block">
        <span>seller </span>
        <Link
          to={`/user/${sellerUsername}?p=1`}
          data-testid="CartAndTransactionItems-item-seller-link"
        >
          <GreenText>{sellerUsername}</GreenText>
        </Link>
      </PlainText>
      {type === itemTypes.CART
        ? listItem.map((item) => (
            <CartItem key={item._id} data={item} isCartLoading={isCartLoading} />
          ))
        : listItem.map((item) => <TransactionAndOrderProdItem key={item._id} data={item} />)}
    </SC.SingleSeller>
  ));

  return sellerList;
};

CartAndTransactionItems.defaultProps = {
  isCartLoading: undefined,
};

CartAndTransactionItems.propTypes = {
  items: PropTypes.arrayOf(PropTypes.object).isRequired,
  type: PropTypes.oneOf([itemTypes.TRANSACTION, itemTypes.CART]).isRequired,
  isCartLoading: PropTypes.bool,
};

export default CartAndTransactionItems;
