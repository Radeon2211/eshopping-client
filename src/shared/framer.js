import theme from '../styled/theme';

export const stepFormVariants = {
  hidden: {
    opacity: 0,
  },
  visible: {
    opacity: 1,
    transition: { duration: theme.durations.level1, delay: theme.durations.level1 },
  },
};

export const backdropVariants = {
  hidden: {
    opacity: 0,
    pointerEvents: 'none',
  },
  visible: {
    opacity: 1,
    pointerEvents: 'initial',
    transition: { duration: theme.durations.level3 },
  },
};

export const modalVariants = {
  hidden: {
    opacity: 0,
    pointerEvents: 'none',
  },
  visible: {
    opacity: 1,
    transition: { delay: theme.durations.level2 },
    pointerEvents: 'initial',
  },
};

export const messageBoxVariants = {
  hidden: {
    pointerEvents: 'none',
    y: '-150%',
  },
  visible: {
    transition: { type: 'spring', duration: theme.durations.level2, stiffness: 100 },
    pointerEvents: 'initial',
    y: 0,
  },
};

export const dropdownVariants = {
  hidden: {
    opacity: 0,
    pointerEvents: 'none',
    y: '-10%',
  },
  visible: {
    opacity: 1,
    pointerEvents: 'initial',
    y: '0',
  },
};
