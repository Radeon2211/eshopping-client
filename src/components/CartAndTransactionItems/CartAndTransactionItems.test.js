import React from 'react';
import { render, cleanup, screen, fireEvent } from '@testing-library/react';
import { Router } from 'react-router-dom';
import { ThemeProvider } from 'styled-components';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import CartAndTransactionItems from './CartAndTransactionItems';
import theme from '../../styled/theme';
import {
  checkProps,
  createCartItem,
  createTransactionAndOrderProdItem,
} from '../../shared/testUtility/testUtility';
import { itemTypes } from '../../shared/constants';

const mockStore = configureMockStore([thunk]);

const defaultStore = mockStore({});

const setUp = (items, type) => {
  const history = {
    listen: jest.fn(),
    createHref: jest.fn(),
    location: { pathname: '/cart' },
    push: jest.fn(),
  };

  return {
    ...render(
      <Router history={history}>
        <Provider store={defaultStore}>
          <ThemeProvider theme={theme}>
            <CartAndTransactionItems items={items} type={type} isCartLoading={false} />
          </ThemeProvider>
        </Provider>
      </Router>,
    ),
    history,
  };
};

afterEach(cleanup);

describe('<CartAndTransactionItems />', () => {
  describe('check prop types', () => {
    it('should NOT throw a warning if type is CART', () => {
      const props = {
        items: [createTransactionAndOrderProdItem()],
        type: itemTypes.CART,
        isCartLoading: false,
      };
      expect(checkProps(CartAndTransactionItems, props)).toBeUndefined();
    });

    it('should NOT throw a warning if type is TRANSACTION', () => {
      const props = {
        items: [createTransactionAndOrderProdItem()],
        type: itemTypes.TRANSACTION,
        isCartLoading: false,
      };
      expect(checkProps(CartAndTransactionItems, props)).toBeUndefined();
    });

    it('should throw a warning if type is ORDER', () => {
      const props = {
        items: [createTransactionAndOrderProdItem()],
        type: itemTypes.ORDER,
        isCartLoading: false,
      };
      expect(checkProps(CartAndTransactionItems, props)).not.toBe(null);
    });

    it('should throw a warning if props are empty', () => {
      expect(checkProps(CartAndTransactionItems, {})).not.toBe(null);
    });
  });

  describe('check how renders', () => {
    it('should render everything with one seller and one cart item', () => {
      const { asFragment } = setUp(
        [
          createCartItem({
            sellerUsername: 'user1',
          }),
        ],
        itemTypes.CART,
      );
      expect(asFragment()).toMatchSnapshot();
    });

    it('should render everything with two sellers, each with one transaction item', () => {
      const { asFragment } = setUp(
        [
          createTransactionAndOrderProdItem({
            productId: 'p1',
            sellerUsername: 'user1',
          }),
          createTransactionAndOrderProdItem({
            productId: 'p2',
            sellerUsername: 'user2',
          }),
        ],
        itemTypes.TRANSACTION,
      );
      expect(asFragment()).toMatchSnapshot();
    });

    it('should call push with correct path after click on user link', () => {
      const { history } = setUp(
        [
          createCartItem({
            sellerUsername: 'user1',
          }),
        ],
        itemTypes.CART,
      );

      fireEvent.click(screen.getByTestId('CartAndTransactionItems-item-seller-link'));
      expect(history.push).toHaveBeenCalledWith('/user/user1?p=1');
    });
  });
});
