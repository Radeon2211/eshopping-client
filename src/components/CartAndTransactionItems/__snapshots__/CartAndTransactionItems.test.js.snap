// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CartAndTransactionItems /> Check how renders Should render everything with one seller and one cart item 1`] = `
<DocumentFragment>
  <div
    class="sc-crrsfI fvpbQF"
  >
    <span
      class="sc-jSgupP dapyFi"
      color=""
      display="block"
      overflow="visible"
      size="3"
      spacing="0px"
      transform=""
    >
      <span>
        seller 
      </span>
      <a
        data-testid="CartAndTransactionItems-seller-link"
      >
        <span
          class="sc-jrAGrp cfzRYv"
        >
          user1
        </span>
      </a>
    </span>
    <div
      class="sc-gKsewC kNJdCK"
      data-testid="CartItem"
    >
      <div
        class="photo-and-name"
      >
        <a
          data-testid="CartItem-product-link-photo"
        >
          <div
            class="sc-bdfBwQ cSJPGs sc-eCssSg gJTqye"
            data-testid="ProductThumbnail"
            direction="row"
            height="7"
            spacing=""
            width="7"
            wrap="nowrap"
          >
            <img
              alt="product name"
              data-testid="ProductThumbnail-img"
              src="no-photo.png"
            />
          </div>
        </a>
        <span
          class="name"
        >
          <a
            data-testid="CartItem-product-link-name"
          >
            product name
          </a>
        </span>
      </div>
      <div
        class="mobile-lower-row"
      >
        <div
          class="choose-quantity-box"
        >
          <div
            class="sc-pFZIQ hgBHyW"
          >
            <button
              class="button minus"
              data-testid="ChooseQuantity-minus-btn"
              disabled=""
              type="button"
            >
              <div
                class="sc-fubCfw iuWfXe"
                color="#DADFE1"
                data-testid="ChooseQuantity-minus-icon"
              >
                <svg>
                  minus.svg
                </svg>
              </div>
            </button>
            <input
              class="sc-iBPRYJ eTbKPo"
              data-testid="NumberInput-quantity"
              name="quantity"
              type="number"
              value="1"
            />
            <button
              class="button plus"
              data-testid="ChooseQuantity-plus-btn"
              type="button"
            >
              <div
                class="sc-fubCfw irUaNG"
                color=""
                data-testid="ChooseQuantity-plus-icon"
              >
                <svg>
                  plus.svg
                </svg>
              </div>
            </button>
          </div>
          <span
            class="sc-jSgupP jgbVDP"
            color=""
            data-testid="CartItem-available-quantity"
            display="inline-block"
            overflow="visible"
            size="2"
            spacing="0px"
            transform=""
          >
            of 2
          </span>
        </div>
        <div
          class="price-box"
        >
          <span
            class="sc-jSgupP esalpy"
            color=""
            display="inline-block"
            overflow="visible"
            size="5"
            spacing="0px"
            transform=""
          >
            $2
          </span>
        </div>
      </div>
      <div
        class="sc-fubCfw eICEUr remove-icon"
        color="#F22613"
      >
        <svg>
          trash.svg
        </svg>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<CartAndTransactionItems /> Check how renders Should render everything with two sellers, each with one transaction item 1`] = `
<DocumentFragment>
  <div
    class="sc-crrsfI fvpbQF"
  >
    <span
      class="sc-jSgupP dapyFi"
      color=""
      display="block"
      overflow="visible"
      size="3"
      spacing="0px"
      transform=""
    >
      <span>
        seller 
      </span>
      <a
        data-testid="CartAndTransactionItems-seller-link"
      >
        <span
          class="sc-jrAGrp cfzRYv"
        >
          user1
        </span>
      </a>
    </span>
    <div
      class="sc-bdfBwQ ZqsAz sc-gsTCUz inazTm"
      data-testid="TransactionAndOrderProdItem"
      direction="row"
      spacing="2"
      width="100%"
      wrap="nowrap"
    >
      <a
        data-testid="TransactionAndOrderProdItem-product-link-photo"
      >
        <div
          class="sc-bdfBwQ igmTYO sc-eCssSg eawKUo"
          data-testid="ProductThumbnail"
          direction="row"
          height="5"
          spacing=""
          width="5"
          wrap="nowrap"
        >
          <img
            alt="product name"
            data-testid="ProductThumbnail-img"
            src="no-photo.png"
          />
        </div>
      </a>
      <div
        class="sc-dlfnbm cjrVjg"
      >
        <span
          class="name"
        >
          <a
            data-testid="TransactionAndOrderProdItem-product-link-name"
          >
            product name
          </a>
        </span>
        <div
          class="sc-bdfBwQ lgjdug sc-hKgILt huHHva"
          direction="row"
          spacing=""
          width="100%"
          wrap="nowrap"
        >
          <span
            class="sc-jSgupP jUhfHv"
            color="#777"
            display="inline-block"
            overflow="visible"
            size="2"
            spacing="0px"
            transform=""
          >
            1 x $2
          </span>
          <span
            class="sc-jSgupP jiAuFg"
            color=""
            display="inline-block"
            overflow="visible"
            size="4"
            spacing="0px"
            transform=""
          >
            $2
          </span>
        </div>
      </div>
    </div>
  </div>
  <div
    class="sc-crrsfI fvpbQF"
  >
    <span
      class="sc-jSgupP dapyFi"
      color=""
      display="block"
      overflow="visible"
      size="3"
      spacing="0px"
      transform=""
    >
      <span>
        seller 
      </span>
      <a
        data-testid="CartAndTransactionItems-seller-link"
      >
        <span
          class="sc-jrAGrp cfzRYv"
        >
          user2
        </span>
      </a>
    </span>
    <div
      class="sc-bdfBwQ ZqsAz sc-gsTCUz inazTm"
      data-testid="TransactionAndOrderProdItem"
      direction="row"
      spacing="2"
      width="100%"
      wrap="nowrap"
    >
      <a
        data-testid="TransactionAndOrderProdItem-product-link-photo"
      >
        <div
          class="sc-bdfBwQ igmTYO sc-eCssSg eawKUo"
          data-testid="ProductThumbnail"
          direction="row"
          height="5"
          spacing=""
          width="5"
          wrap="nowrap"
        >
          <img
            alt="product name"
            data-testid="ProductThumbnail-img"
            src="no-photo.png"
          />
        </div>
      </a>
      <div
        class="sc-dlfnbm cjrVjg"
      >
        <span
          class="name"
        >
          <a
            data-testid="TransactionAndOrderProdItem-product-link-name"
          >
            product name
          </a>
        </span>
        <div
          class="sc-bdfBwQ lgjdug sc-hKgILt huHHva"
          direction="row"
          spacing=""
          width="100%"
          wrap="nowrap"
        >
          <span
            class="sc-jSgupP jUhfHv"
            color="#777"
            display="inline-block"
            overflow="visible"
            size="2"
            spacing="0px"
            transform=""
          >
            1 x $2
          </span>
          <span
            class="sc-jSgupP jiAuFg"
            color=""
            display="inline-block"
            overflow="visible"
            size="4"
            spacing="0px"
            transform=""
          >
            $2
          </span>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
